test/dataset.go

```diff
@@ -0,0 +1,258 @@
+	LastInsertId = "lastInsertId"
+	ValueInt     = "valueInt"
+	ValueString  = "valueString"
+	File         = "file"
+)
+
+type (
+	// Message expected message or actual message for integration test
+	Message struct {
+		Kind     string   `yaml:"kind"` // dataset, expected
+		MetaData MetaData `yaml:"metadata"`
+		Data     []*Data  `yaml:"data"`
+	}
+
+	MetaData struct {
+		Tables []*Table `json:"tables"`
+	}
+
+	Data struct {
+		Name  string     `json:"name"`
+		Value [][]string `json:"value"`
+	}
+
+	Table struct {
+		Name    string   `yaml:"name"`
+		Columns []Column `yaml:"columns"`
+	}
+
+	Column struct {
+		Name string `yaml:"name"`
+		Type string `yaml:"type"`
+	}
+)
+
+type (
+	Cases struct {
+		Kind           string  `yaml:"kind"` // dataset, expected
+		ExecCases      []*Case `yaml:"exec_cases"`
+		QueryRowsCases []*Case `yaml:"query_rows_cases"`
+		QueryRowCases  []*Case `yaml:"query_row_cases"`
+	}
+
+	Case struct {
+		SQL            string    `yaml:"sql"`
+		Parameters     string    `yaml:"parameters,omitempty"`
+		Type           string    `yaml:"type"`
+		Sense          []string  `yaml:"sense"`
+		ExpectedResult *Expected `yaml:"expected"`
+	}
+
+	Expected struct {
+		ResultType string `yaml:"resultType"`
+		Value      string `yaml:"value"`
+	}
+)
+
+func (e *Expected) CompareRows(rows *sql.Rows, actual *Message) error {
+	pr := func(t interface{}) (r string) {
+		r = "\\N"
+		switch v := t.(type) {
+		case *sql.NullBool:
+			if v.Valid {
+				r = strconv.FormatBool(v.Bool)
+			}
+		case *sql.NullString:
+			if v.Valid {
+				r = v.String
+			}
+		case *sql.NullInt64:
+			if v.Valid {
+				r = fmt.Sprintf("%6d", v.Int64)
+			}
+		case *sql.NullFloat64:
+			if v.Valid {
+				r = fmt.Sprintf("%.2f", v.Float64)
+			}
+		case *time.Time:
+			if v.Year() > 1900 {
+				r = v.Format("_2 Jan 2006")
+			}
+		default:
+			r = fmt.Sprintf("%#v", t)
+		}
+		return
+	}
+
+	c, _ := rows.Columns()
+	n := len(c)
+	field := make([]interface{}, 0, n)
+	for i := 0; i < n; i++ {
+		field = append(field, new(sql.NullString))
+	}
+
+	var converts [][]string
+
+	for rows.Next() {
+		if err := rows.Scan(field...); err != nil {
+			return err
+		}
+		row := make([]string, 0, n)
+		for i := 0; i < n; i++ {
+			col := pr(field[i])
+			row = append(row, col)
+		}
+		converts = append(converts, row)
+	}
+
+	actualMap := make(map[string][]map[string]string, 10) // table:key:value
+	for _, v := range actual.Data {
+		tb := actualMap[v.Name]
+		if tb == nil {
+			actualMap[v.Name] = make([]map[string]string, 0, 1000)
+		}
+		for _, rows := range v.Value {
+			val := make(map[string]string, len(rows))
+			for i, row := range rows {
+				val[actual.MetaData.Tables[0].Columns[i].Name] = row
+				actualMap[v.Name] = append(actualMap[v.Name], val)
+			}
+		}
+	}
+
+	for _, rows := range converts {
+		foundFlag := false
+		// for loop actual data
+		for _, actualRows := range actualMap {
+			for _, actualRow := range actualRows {
+				if actualRow[c[0]] == rows[0] {
```

Comment:
Maybe refactor a new method for data compare, reduce complexity.

Reply:
addressed